devito/finite_differences/finite_difference.py

```diff
@@ -258,34 +279,19 @@ def generic_derivative(expr, dim, fd_order, deriv_order, stagger=None):
 
     diff = dim.spacing
 
-    if stagger == left or not expr.is_Staggered:
-        off = -.5
-    elif stagger == right:
-        off = .5
-    else:
-        off = 0
-
-    if expr.is_Staggered:
-        indices = list(set([(dim + int(i+.5+off) * dim.spacing)
-                            for i in range(-fd_order//2, fd_order//2)]))
-        x0 = (dim + off*diff)
-        if fd_order < 2:
-            indices = [dim + diff, dim] if stagger == right else [dim - diff, dim]
+    indices, x0 = generate_indices(expr, dim, diff, fd_order, stagger=stagger)
 
+    if symbolic:
+        c = symbolic_weights(expr, deriv_order, indices, x0)
     else:
-        indices = [(dim + i * dim.spacing) for i in range(-fd_order//2, fd_order//2 + 1)]
-        x0 = dim
-        if fd_order < 2:
-            indices = [dim, dim + diff]
-
-    c = finite_diff_weights(deriv_order, indices, x0)[-1][-1]
+        c = finite_diff_weights(deriv_order, indices, x0)[-1][-1]
 
     deriv = 0
     all_dims = tuple(set((dim, ) +
                      tuple([i for i in expr.indices if i.root == dim])))
     for i in range(0, len(indices)):
-            subs = dict([(d, indices[i].subs({dim: d})) for d in all_dims])
-            deriv += expr.subs(subs) * c[i]
+        subs = dict([(d, indices[i].subs({dim: d})) for d in all_dims])
```

Comment:
probably no need to wrap within a [...] (it might save some time to pass a ``generator`` directly to ``dict``)

Reply:
Redundant brackets removed.